### PR TITLE
DAPP-1552-blocks/Separator

### LINKED ISSUE #1552: [DApp - DS] Create Separator Component

### PR DESCRIPTION
**Pull request by kalashshah**
### Ticket Number

- [Ticket Number](#1552 )
- Fixes #1552 

### Description

- **Problem/Feature**:
- Feature: Introduced a new Separator component in design system. The Separator component accepts the orientation as a prop and renders it accordingly.
- Additional styles can be passed as props and supports an additional prop named `css` which supports additional styling similar to styled components taking reference from the `typography` component.

### Type of Change

- [x] New feature

### Checklist

- [ ] **Quick PR**: Is this a quick PR? Can be approved before finishing a coffee.
  - [ ] Quick PR label added
- [ ] **Not Merge Ready**: Is this PR dependent on some other PR/tasks and not ready to be merged right now.
  - [ ] DO NOT Merge PR label added

### Frontend Guidelines

- [ ] Followed frontend guidelines as per the [Guidelines Notion Doc](https://www.notion.so/pushprotocol/Frontend-dApp-Guidelines-1d7806ae3d9e4569a340b563dcd0536c)

### Build & Testing

- [x] No errors in the build terminal
- [x] Engineer has tested the changes on their local environment
- [ ] Engineer has tested the changes on deploy preview

### Screenshots/Video with Explanation

Screenshot 2024-06-04 at 5 29 22 PM

### Review & Approvals

- [x] Self-review completed
- [ ] Code review by at least one other engineer
- [ ] Documentation updates if applicable

### CODE REVIEW

File: src/blocks/separator/Separator.types.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  /* Sets min-height css property */
+  minHeight?: ResponsiveProp<string>;
+  /* Sets max-width css property */
+  maxWidth?: ResponsiveProp<string>;
+  /* Sets min-width css property */
+  minWidth?: ResponsiveProp<string>;
+  /* Sets padding css property */
+  padding?: ResponsiveProp<SpaceType>;
+  /* Sets width css property */
+  width?: ResponsiveProp<string>;
```
Comment (rohitmalhotra1420) [inline src/blocks/separator/Separator.types.ts:21]: I don't think we 

- Don't need maxHeight, minHeight css props.  Separator component should be automatically take full available height of the parent component unless given a specific height.
- Don't need maxWidth, minWidth css props.  Separator component should be automatically take full available width of the parent component unless given a specific width.

File: src/blocks/separator/Separator.types.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  /* Sets border css property */
+  border?: string;
+  /* Sets border-radius css property */
+  borderRadius?: string;
+  /* Sets background-color css property */
+  backgroundColor?: BlocksColors;
+  /* Sets position css property */
+  position?: CSSProperties['position'];
+  /* Sets box-shadow css property */
+  boxShadow?: string;
```
Comment (rohitmalhotra1420) [inline src/blocks/separator/Separator.types.ts:34]: None of these props are required. Separator doesn't needs these css properties.

File: src/blocks/separator/Separator.types.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+};
+
+export type SeparatorCSSProps = SeparatorResponsiveProps & SeparatorNonResponsiveProps;
+
+export type SeparatorComponentProps = {
+  /* Additional prop from styled components to apply custom css to Separator */
+  css?: FlattenSimpleInterpolation;
+  /* Orientation of the separator */
+  orientation?: 'horizontal' | 'vertical';
```
Comment (rohitmalhotra1420) [inline src/blocks/separator/Separator.types.ts:43]: This prop is passed down to the html tag. We need to add orientation into `separatorCSSPropsKeys` array.

File: src/blocks/separator/Separator.utils.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  { propName: 'padding', prop: props.padding },
+  { propName: 'width', prop: props.width },
+];
+
+export const getSeparatorResponsiveCSS = (props: SeparatorResponsiveProps) => {
+  const data = getSeparatorResponsiveCSSProperties(props);
+  return getResponsiveCSS(data);
+};
+
+export const separatorCSSPropsKeys: (keyof SeparatorCSSProps)[] = [
```
Comment (rohitmalhotra1420) [inline src/blocks/separator/Separator.utils.ts:22]: Let's move this to Separator.constants.ts. Could you also do the same for `boxCSSPropsKeys`.

File: src/blocks/separator/Separator.tsx [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  border-radius: ${({ borderRadius }) => borderRadius};
+  border: ${({ border }) => border};
+  position: ${({ position }) => position};
+
+  /* Extra CSS prop */
+  ${({ css }) => css || ''}
+`;
+
+const Separator = forwardRef<HTMLDivElement, SeparatorProps>(
+  ({ backgroundColor = 'lightGray', orientation = 'horizontal', ...props }, ref) => {
```
Comment (rohitmalhotra1420) [inline src/blocks/separator/Separator.tsx:34]: We don't need backgroundColor: separator needs to be single color through out the app.

File: src/blocks/separator/Separator.tsx [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+export type SeparatorProps = SeparatorCSSProps & SeparatorComponentProps & BlockWithoutStyleProp<HTMLDivElement>;
+
+const StyledSeparator = styled.div.withConfig({
+  shouldForwardProp: (prop, defaultValidatorFn) =>
+    !separatorCSSPropsKeys.includes(prop as keyof SeparatorCSSProps) && defaultValidatorFn(prop),
+})<SeparatorProps>`
+  /* Initial values */
+  width: ${({ width, orientation }) => width || (orientation === 'horizontal' ? 'auto' : '1px')};
+  height: ${({ height, orientation }) => height || (orientation === 'horizontal' ? '1px' : 'auto')};
```
Comment (rohitmalhotra1420) [inline src/blocks/separator/Separator.tsx:17]: When the orientation is horizontal, the separator should automatically take the full width of the parent container.

When the orientation is vertical, the separator should automatically take the full height of the parent container.

until a width or height is passed as a prop from outside.

Right now I have to pass width/height:100% everytime if I wan't to render a full length container.

File: src/blocks/separator/Separator.tsx [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  box-shadow: ${({ boxShadow }) => boxShadow};
+  border-radius: ${({ borderRadius }) => borderRadius};
+  border: ${({ border }) => border};
+  position: ${({ position }) => position};
+
+  /* Extra CSS prop */
+  ${({ css }) => css || ''}
+`;
+
+const Separator = forwardRef<HTMLDivElement, SeparatorProps>(
```
Comment (rohitmalhotra1420) [inline src/blocks/separator/Separator.tsx:33]: We don't need forwardRef for the Separator component. Can't think of any use case so let's remove this as well.

**Review by rohitmalhotra1420 — CHANGES_REQUESTED** [review]

File: src/blocks/separator/Separator.tsx [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  ({ backgroundColor = 'lightGray', orientation = 'horizontal', ...props }, ref) => {
+    return (
+      <StyledSeparator
+        ref={ref}
```
Comment (rohitmalhotra1420) [inline src/blocks/separator/Separator.tsx:37]:
```suggestion
        role="separator"
        ref={ref}
```

File: src/blocks/separator/Separator.types.ts (hunk 2 above)
Comment (kalashshah) [reply src/blocks/separator/Separator.types.ts:34]: It is possible that we might need to add a `borderRadius` and a `backgroundColor` to the separator right? Agree that the rest have an extremely rare use case.

Comment (rohitmalhotra1420) [reply src/blocks/separator/Separator.types.ts:34]: No we don't. The whole idea is to make is consistent and use a standard version. It should be same through out the app

File: src/blocks/separator/Separator.types.ts [hunk 9]
```diff
@@ -1,61 +1,34 @@
...
-  maxWidth?: ResponsiveProp<string>;
-  /* Sets min-width css property */
-  minWidth?: ResponsiveProp<string>;
-  /* Sets padding css property */
-  padding?: ResponsiveProp<SpaceType>;
   /* Sets width css property */
   width?: ResponsiveProp<string>;
 };
 
 export type SeparatorNonResponsiveProps = {
```
Comment (rohitmalhotra1420) [inline src/blocks/separator/Separator.types.ts:13]: Let's remove SeparatorNonResponsiveProps as it's not required and combine it's props into SeparatorComponentProps

File: src/blocks/separator/Separator.types.ts [hunk 10]
```diff
@@ -1,61 +1,34 @@
...
+export type SeparatorResponsiveCSSProperties = 'height' | 'margin' | 'width';
 
 export type SeparatorResponsivePropValues = ValueOf<SeparatorResponsiveProps>;
 
 export type SeparatorResponsiveCSSPropertiesData = {
   propName: SeparatorResponsiveCSSProperties;
   prop: SeparatorResponsivePropValues;
 };
+
+export const separatorCSSPropsKeys: (keyof SeparatorCSSProps)[] = ['height', 'margin', 'width', 'orientation'];
```
Comment (rohitmalhotra1420) [inline src/blocks/separator/Separator.types.ts:34]: This needs to go to Separator.constants.ts

Comment (rohitmalhotra1420) [inline src/blocks/separator/Separator.types.ts:34]: `export const separatorRestrictedPropsKeys: (keyof SeparatorProps)[] = ['height', 'margin', 'width', 'orientation'];
`

File: src/blocks/separator/Separator.types.ts [hunk 11]
```diff
@@ -1,61 +1,34 @@
...
-  backgroundColor?: BlocksColors;
-  /* Sets position css property */
-  position?: CSSProperties['position'];
-  /* Sets box-shadow css property */
-  boxShadow?: string;
+  /* Orientation of the separator */
+  orientation?: 'horizontal' | 'vertical';
 };
 
 export type SeparatorCSSProps = SeparatorResponsiveProps & SeparatorNonResponsiveProps;
```
Comment (rohitmalhotra1420) [inline src/blocks/separator/Separator.types.ts:18]: SeparatorCSSProps can be removed

File: src/blocks/separator/Separator.tsx [hunk 12]
```diff
@@ -1,10 +1,10 @@
-import { forwardRef } from 'react';
+import React from 'react';
```
Comment (rohitmalhotra1420) [inline src/blocks/separator/Separator.tsx:1]: Let's not use direct react imports.

`
import { FC } from 'react';
`

File: src/blocks/separator/Separator.tsx [hunk 13]
```diff
@@ -13,34 +13,28 @@ const StyledSeparator = styled.div.withConfig({
...
-    return (
-      <StyledSeparator
-        ref={ref}
-        {...props}
-        {...{ orientation, backgroundColor }}
-      />
-    );
-  }
-);
+const Separator: React.FC<SeparatorProps> = ({ orientation = 'horizontal', ...rest }) => {
```
Comment (rohitmalhotra1420) [inline src/blocks/separator/Separator.tsx:29]: `const Separator: FC<SeparatorProps> = ({ orientation = 'horizontal', ...rest }) => {`

File: src/blocks/box/Box.types.ts [hunk 14]
```diff
@@ -77,3 +77,25 @@ export type BoxResponsiveCSSProperties =
 export type BoxResponsivePropValues = ValueOf<BoxResponsiveProps>;
 
 export type BoxResponsiveCSSPropertiesData = { propName: BoxResponsiveCSSProperties; prop: BoxResponsivePropValues };
+
+export const boxCSSPropsKeys: (keyof BoxCSSProps)[] = [
```
Comment (rohitmalhotra1420) [inline src/blocks/box/Box.types.ts:81]: Move it to box.constants.ts

**Review by rohitmalhotra1420 — CHANGES_REQUESTED** [review]

**Review by rohitmalhotra1420 — APPROVED** [review]
Look good. Works fine. 

Nicely done @kalashshah

**Review by corlard3y — APPROVED** [review]

**Pull request merged**